src/relay/backend/te_compiler.cc

```diff
@@ -134,41 +144,78 @@ class TECompilerImpl : public TECompilerNode {
+                              function_node->type_params, /*attrs=*/{}, function_node->span);
+            // Mark function as 'extern' using the "ExternalSymbol" attribute.
+            function = WithAttr(std::move(function), attr::kExternalSymbol, kv2.first->name_hint);
+            module->Add(kv2.first, function);
+          }
+        }
+      }
+    }
+  }
+
   Array<tvm::runtime::Module> LowerExternalFunctions() {
     Array<tvm::runtime::Module> ret;
-    std::unordered_map<std::string, std::string> cached_symbol;
     std::vector<CCacheKey> cached_ext_funcs;
 
     for (const auto& it : cache_) {
       auto src_func = it.first->source_func;
       ICHECK(src_func.defined());
-      if (src_func->GetAttr<String>(attr::kCompiler).defined()) {
-        auto code_gen = src_func->GetAttr<String>(attr::kCompiler);
-        std::string code_gen_name = code_gen.value();
+      Optional<String> opt_compiler = src_func->GetAttr<String>(attr::kCompiler);
+      if (opt_compiler.defined()) {
+        Optional<String> opt_symbol_name = src_func->GetAttr<String>(tvm::attr::kGlobalSymbol);
+        ICHECK(opt_symbol_name.defined()) << "No external symbol is set for:" << std::endl
+                                          << PrettyPrint(src_func);
+        VLOG(1) << "using external codegen '" << opt_compiler.value() << "' for name '"
+                << opt_symbol_name.value() << "' and function:" << std::endl
+                << PrettyPrint(src_func);
         cached_ext_funcs.push_back(it.first);
 
-        auto symbol_name = src_func->GetAttr<String>(tvm::attr::kGlobalSymbol);
-        ICHECK(symbol_name.defined()) << "No external symbol is set for:\n"
-                                      << AsText(src_func, false);
-
-        std::string sn = symbol_name.value();
-        if (cached_symbol.count(sn)) {
-          cached_symbol[sn] = code_gen_name;
-        } else {
-          ICHECK_NE(sn, code_gen_name)
-              << "Found duplicated symbol: " << sn << " for: " << code_gen_name;
-        }
-
-        std::string ext_name = "relay.ext." + code_gen_name;
+        std::string ext_name = "relay.ext." + opt_compiler.value();
         auto pf = tvm::runtime::Registry::Get(ext_name);
         ICHECK(pf) << "Failed to find the codegen tool for " << ext_name;
         // No need to keep compiler attribute at this point, functions have been
         // extracted for specific codegen.
         src_func = WithAttr(std::move(src_func), attr::kCompiler, NullValue<ObjectRef>());
+        VLOG_CONTEXT << ext_name;
         runtime::Module ext_mod = (*pf)(src_func);
-
-        ICHECK(ext_mod.defined()) << "No external runtime is generated.";
-        ret.push_back(ext_mod);
+        if (ext_mod.defined()) {
+          if (ext_mod->GetFunction(opt_symbol_name.value(), /*query_imports=*/true) == nullptr) {
+            // It's possible the codegen yielded C or C++ tracked separately and thus the
+            // returned runtime module can be empty.
```

Comment:
Is there a way to check that the the codegen made C/C++ that is tracked separately? Would be nice to have a check for that here. Otherwise, maybe add some of the comment to the VLOG message?

Reply:
Or maybe log a warning.

Reply:
I asked our ARM friends about this since it was also bugging me. It turns out the runtime::Module will be an EthosUModule which does not support GetFunction for the compiled functions, but *does* support GetFunction for the meta 'get_func_names' which would yield a list of those names. However that's not a standard interface te_compiler can depend on. Better would be HasFunction or something on runtime::Module but I'm leaving that for a rain(ier) day.